File: src/main/java/me/austin/queer/module/gui/clickgui/components/ClickGuiScreen.java
package me.austin.queer.module.gui.clickgui.components;

import java.util.List;

import me.austin.queer.module.hacks.Category;
import me.austin.queer.module.hacks.Hacks;
import me.austin.queer.module.setting.Setting;
import me.austin.queer.util.ScreenHelper;

import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.text.LiteralText;

public class ClickGuiScreen extends Screen implements ScreenHelper {
	public static List<CategoryFrame> frames;
	public static boolean shouldCloseOnEsc;
	public Hacks hacks;
	
	public ClickGuiScreen(Hacks hacks, Setting<Integer> backgroundRed, Setting<Integer> backgroundGreen, Setting<Integer> backgroundBlue) {
		super(new LiteralText("ClickGuiScreen"));

		this.hacks = hacks;
		int offset = 20;
		for (Category category : Category.values()) {
			frames.add(new CategoryFrame(category, (20 + offset) * wpixel, 20 * hpixel, this));
			offset += 60;
		}
	}

	@Override
	public void render(MatrixStack matrices, int mousex, int mousey, float delta) {
		for (CategoryFrame frame : frames) {
			frame.render(matrices, mc.textRenderer, frame.x, frame.y);
		}
	}
	
	@Override
	public boolean mouseClicked(double mousex, double mousey, int button) {
		for (CategoryFrame frame : frames) {
			if (ScreenHelper.clickCheck(mousex, mousey, frame.x, frame.y, frame.width, 20)) {
				frame.mouseClicked(mousex, mousey, button);
				return true;
			}
		}
		return false;
	}

	@Override 
	public boolean mouseReleased(double mousex, double mousey, int button) {
		for (CategoryFrame frame : frames) {
			if (frame.dragging) {
				frame.dragging = false;
				return true;
			}
		}
		return false;
	} 
	
	@Override
	public boolean isPauseScreen() {
		return false;
	}

	@Override
	public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
		frames.forEach(frame -> {
			frame.keyPressed(keyCode, scanCode, modifiers);
		}); 
		return true;
	}

	@Override
	public boolean shouldCloseOnEsc() {
		return shouldCloseOnEsc;
	}

	@Override
	public void renderBackground(MatrixStack matrices, int vOffset) {
		return;
	}
}

File: src/main/kotlin/trans/rights/impl/gui/ClickGuiScreen.kt
package trans.rights.impl.gui

import net.minecraft.client.gui.DrawableHelper
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.text.Text
import trans.rights.BasicEventManager
import trans.rights.Queer
import trans.rights.api.Wrapper
import trans.rights.api.gui.components.Frame
import trans.rights.api.gui.components.buttons.Button
import trans.rights.api.hack.Hack
import trans.rights.api.hack.HackManager
import trans.rights.events.KeyEvent
import me.austin.rush.listener
import java.awt.Color
import java.nio.file.Files
import java.nio.file.Path

object ClickGuiScreen : Screen(Text.of(Queer.NAME)), Wrapper, trans.rights.api.Manager<Frame<*>, List<Frame<*>>> {
    override val values: List<Frame<*>>

    private val file: Path = Path.of("${Queer.mainDirectory}/clickguiscreen.json")

    //     private var key: Int = file.fromJson().get("key").asInt
    private var shouldCloseOnEsc = true

    private var key = 'y'

    private val keyListener = listener<KeyEvent>({
        if (it.key == this.key.code) {
            minecraft.setScreen(ClickGuiScreen)
            it.cancel()
        }
    }, Integer.MAX_VALUE)

    init {
        var offset = 0

        values = listOf(object : Frame<Hack>(20, 20, 60, HackManager.values.size * 20, HackManager.values.map {
            offset += 20

            object : Button<Hack>(20, offset, 60, 20, it) {
                override fun render(stack: MatrixStack) {
                    DrawableHelper.fill(
                        stack,
                        this.xPos,
                        this.yPos,
                        this.width + this.xPos,
                        this.height + this.yPos,
                        Color.LIGHT_GRAY.rgb
                    )
                }
            }
        }) {
            override fun render(stack: MatrixStack) {
                DrawableHelper.fill(
                    stack,
                    this.xPos,
                    this.yPos,
                    this.width + this.xPos,
                    this.height + this.yPos,
                    Color.LIGHT_GRAY.rgb
                )

                for (button in this.children) button.render(stack)
            }
        })
    }

    override fun render(matrices: MatrixStack, mouseX: Int, mouseY: Int, delta: Float) {
        for (frame in values) frame.render(matrices)
    }

    override fun shouldPause() = false

    override fun shouldCloseOnEsc() = shouldCloseOnEsc

    override fun load() {
        if (!Files.exists(file)) Files.createFile(file)

        BasicEventManager.register(keyListener)
    }

    override fun unload() {
        BasicEventManager.unregister(keyListener)
    }
}